### DISCUSSION TITLE
Implementing code-based email verification flow; issues _without_ verify_account_grace_period enabled

### DISCUSSION THREAD

**Discussion by moneill** [janko/rodauth-rails#322] [Q&A]
Hey there! I'm in the process of implementing code-based email verification following the discussion here: https://groups.google.com/g/rodauth/c/eBQem6q3Ne0/m/4jsJK7EZAwAJ. This is the rough account registration & verification flow I am trying to create:

1/ User creates a new account via /create-account and user is immediately redirected to /verify-account
2/ Verification email (containing 6 digit verification code) gets sent to user
3/ User enters the code from their email on /verify-account
4/ User is verified and redirected

With the `verify_account_grace_period` feature enabled, the above works as expected so long as the verification is completed within the grace period.

However, if the grace period has elapsed, or if `verify_account_grace_period` is disabled (my preference), the following exception occurs after a user enters their verification code:

```
undefined method `verification_key' for nil


Extracted source (around line #199):
              

    account_from_verify_account_key do |key|
      unless timing_safe_eql?(key, rails_account.verification_key.key)
        verify_account_key_error "Invalid verification key."
      end

Application Trace | Framework Trace | Full Trace
app/misc/rodauth_main.rb:199:in `block (2 levels) in <class:RodauthMain>'
app/misc/rodauth_app.rb:11:in `block in <class:RodauthApp>'
```

I have a barebones repo set up here to reproduce this: https://github.com/moneill/rodauth-debug (see [rodauth_main.rb](https://github.com/moneill/rodauth-debug/blob/main/app/misc/rodauth_main.rb)). My approach is a blend of https://groups.google.com/g/rodauth/c/eBQem6q3Ne0/m/4jsJK7EZAwAJ plus https://github.com/janko/rodauth-rails/discussions/249#discussioncomment-7722233 for redirecting to the originally-requested page.

Any thoughts on what might I might be doing wrong? Appreciate any guidance folks might be able to offer!

**Comment by janko** [comment]
If you're not using `verify_account_grace_period` and have `verify_account_autologin? false`, then there will be no account logged in at the time of verification, so `rails_account` will return `nil` (that method returns the currently logged in account).

The following should probably work, though it's probably vulnerable to timing attacks:

```rb
account_from_verify_account_key do |code|
  verification_key = Account::VerificationKey.find_by(code: code)

  if verification_key.nil?
    verify_account_code_error "Invalid verification code."
  elsif verification_key.requested_at <= 24.hours.ago
    verify_account_code_error "This verification code has expired."
  end

  _account_from_id(verification_key.id)
end
```

**Reply by moneill** [reply]
Interestingly, in my setup `verify_account_autologin? true` because I'm not explicitly disabling that option, and so I assumed (perhaps wrongly!) that `rails_account` would be defined at the time of verification. Not sure if that suggests a different possible change here, but I'll give your proposal a shot!

**Reply by janko** [reply]
Actually, even with `verify_account_autologin? true`, the `rails_account` would still be `nil`, because that method searches only verified accounts (it's based on `rodauth.account_from_session`).

BTW, thank you for your sponsorship, it's highly appreciated! ❤️ 🙏🏻

**Reply by moneill** [reply]
So in going through all of this, I *just* realized that this flow is entirely disregarding the traditional (much longer and more complex) verification key and instead using _only_ the 6 digit code to find a user (obvious in retrospect...duh). I don't love that, so I ended up with a hybrid approach that retains the original verification key flow AND adds a 6 digit code. I suspect this is quite a bit more secure and less susceptible to timing attacks. Here's where I have landed, so far...

Add a separate `code` column:
```
class AddCodeToAccountVerificationKeys < ActiveRecord::Migration[7.2]
  def change
    add_column :account_verification_keys, :code, :string
  end
end
```

Then in `rodauth_main.rb`, I altered what's present in my demo repo to the following:

```
account_from_verify_account_key do |key|
  @account = account_from_key(key, account_unverified_status_value) { |id| get_verify_account_key(id) }

  verify_account_code_error "Invalid verification key" if @account.blank?

  unless timing_safe_eql?(param("code"), rails_account.verification_key.code)
    verify_account_code_error "Invalid verification code"
  end

  if rails_account.verification_key.requested_at <= 24.hours.ago
    regenerate_verify_account_code
    verify_account_email_resend
    verify_account_code_error "This verification key has expired. We've sent a new code to #{rails_account.email}."
  end

  @account
end

...

def verify_account_key_insert_hash
  { verify_account_id_column => account_id, verify_account_key_column => verify_account_key_value,
    "code" => random_verify_code }
end
```

And in `verify_account.html.erb`
```
<%= form_with url: rodauth.verify_account_path, method: :post, data: { turbo: false }, class: "w-full max-w-sm" do |form| %>
  <div class="invalid-feedback">
    <% if rodauth.field_error("code") %>
      <%= render "shared/alerts/error" do %>
        <%= rodauth.field_error("code") %>
      <% end %>
    <% end %>
  </div>

  <div class="relative">
    <%= form.number_field :code, placeholder: " ", autocomplete: "one-time-code", class: "block rounded-lg px-4 pb-2.5 pt-5 w-full text-base md:text-sm bg-white border-1 appearance-none focus:outline-none focus:border-dark-navy peer #{rodauth.field_error("code") ? "border-red-800/50 focus:ring-red-600 focus:border-red-600 dark:border-red-800/50 dark:focus:ring-red-400" : "border-neutral-300" }" %>
    <%= form.label :code, "Verification Code", class: "absolute text-base md:text-sm duration-300 transform -translate-y-4 scale-75 top-4 z-10 origin-[0] start-4 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-4 rtl:peer-focus:translate-x-1/4 rtl:peer-focus:left-auto" %>
    <%= form.hidden_field :key, value: params[:key] %>
  </div>
  
  <%= form.submit rodauth.verify_account_button, class: "w-full mt-4 px-8 py-3 cursor-pointer font-semibold text-base md:text-sm rounded-md text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-600 dark:bg-emerald-400 dark:hover:bg-emerald-500 dark:text-gray-900 dark:focus:ring-emerald-400 dark:focus:ring-offset-current" %>
  <p class="text-sm text-center mt-4">
    <%= link_to "Resend Verification Email", rodauth.verify_account_resend_path, class: "underline pl-1" %>
  </p>
<% end %>
```

I need to spend some more time testing, but I think this will work. @janko holler if I'm doing something totally wonky here, but your comments were super helpful in getting me to take a step back. And happy to sponsor -- love this library and Shrine. ❤️

**Reply by janko** [reply]
Yeah, using only a 6-digit code is less secure than tokens, the original code had it because it was copying GitHub's authentication flow that AFAIR doesn't use tokens for account verification. BTW, the first line of the `account_from_verify_account_key` block can be just `super(key)` 😉 

Initially, I didn't see how having a token + code was more secure than just token. However, it could happen that the user visits their verify account link, which Rodauth then stores in the session to prevent token leakage via request referrer, but then aborts the process. Another person on the same computer could then re-open the verify account page, which will work because the token is still stored in the session, and verify someone else's account.

Not sure how serious is that security vulnerability, @jeremyevans what do you think?

**Reply by jeremyevans** [reply]
Certainly, a 6 digit code without token is a significant security vulnerability, unless you enforcing lockout after a small number of attempts.  Otherwise, brute forcing the code is trivial.

**Reply by janko** [reply]
I was referring to the default Rodauth behavior with tokens (no 6-digit codes). When opening the verify account page from the email link, Rodauth will store the token in the session and redirect to the verify account page without the query param, to avoid leaking tokens via the `Referer` header if the user were to click on an external link.

However, that means if the user didn't finish verifying their account, they might not be aware that the verification token is remembered by the browser, an attacker could finish it for them on the same computer. Especially since the account verification token doesn't appear to have an expiration. Though the attacker could probably also look at the browser history as well.

**Reply by janko** [reply]
The 6-digit code *in addition to* the token sounds like it would patch this potential security vulnerability, because it's not enough for the attacker to retrieve the token, they would also have to know the contents of the email.

**Reply by jeremyevans** [reply]
I wouldn't consider the current behavior to be a vulnerability.  Certainly using an additional 6-digit code would add an additional step, and therefore a small amount of security, but it also adds more work for the user, and for most applications I don't think the tradeoffs would be worth it.

Rodauth moves the token from the param to the session, which is the same place it stores the logged in user?  So to say this is a vulnerability would be to say that how the logged in user is stored is also a vulnerability, correct?  If an attacker gets access to a user's browser after they have logged in, you get the same issue, I think.

**Reply by janko** [reply]
For me it's not about the place where token is stored, but the expectations. When the user is logged in, they know they need to sign out if they're on a shared computer. But if they opened the verify account page and leave it, they might not expect that it was persisted somewhere. It's an implementation detail of Rodauth, e.g. Devise won't put anything in the session.

But a trivial browser history search would also uncover the unused verify account token, so I think I was just being paranoid 😄 Anyway, it's good to know that an additional code can easily be added to the account verification flow for people wanting extra security 👍🏻
